File: webapp/frontend/views/pages/article_viewer/article_viewer.py
import os
import tornado.ioloop
import tornado.web

from webapi.modelsDTO.authors import AuthorDTO

from webapi.models.article import Article


class ArticleViewer(tornado.web.RequestHandler):

    def get(self, author_:AuthorDTO = None):
        
        self.render(
            "article_viewer.html", 
            title="Aticle Viewer",
            current_author = AuthorDTO(
                name = "Author name",
                surname = "Author surname",
                profile_image = "",
                miniature = os.path.join("images", "avatar-image.webp")
            ),
            article_tags = ["Python", "Coding", "Algorithms"],
            player = {
                "duration": "6:56",
                "buttons":{
                    "play_button":
                    {
                        "icon1_path": os.path.join('images','cicle-filled-play-icon-2.webp'),
                        "icon2_path": os.path.join('images','cicle-filled-pause-icon-3.webp'),
                        "title": "Click here to start or stop reading",
                        "alt": "play-pause",
                        "on_click": "articlesViewerFunctions.readDocument(this);"
                    },
                    "previous_button":
                    {
                        "icon_path": os.path.join('images','previous-play-icon.webp'),
                        "title": "Click here to move back",
                        "alt": "previous",
                        "on_click": ""
                    },
                    "next_button":
                    {
                        "icon_path": os.path.join('images','next-play-icon.webp'),
                        "title": "Click here ti move next",
                        "alt": "next",
                        "on_click": ""
                    },
                    "speed_button":
                    {
                        "icon_path": "",
                        "title": "Click here to change to speed",
                        "alt": "speed",
                        "on_click": ""
                    },
                    "voice_button":
                    {
                        "icon_path": os.path.join('images','voice-recognition-icon.webp'),
                        "title": "Click here to change the voice",
                        "alt": "voice",
                        "on_click": ""
                    }
                }
            },
            author_actions_bar_button_details = {
                'back': {
                    "name": "",
                    "on_click":"articlesViewerFunctions.previous(this);",
                    "icon_path":os.path.join("images", "left-arrow-icon.webp"),
                    "title": "Click here to back to editorial page",
                    "alt":"back"
                },
                'edit': {
                    "name": "Edit",
                    "on_click":"articlesViewerFunctions.openEditor(this)",
                    "icon_path":"",
                    "title": "Click here to edit the article",
                    "alt":""
                },
                'delete':{
                    "name": "Delete",
                    "on_click":"",
                    "icon_path":"",
                    "title": "Click here to delete the article",
                    "alt":""
                }
            },
            dark_searh_icon_path = os.path.join("images", "dark-search-icon.webp"),
            categories = ["Data science"]
        )

File: webapp/frontend/views/pages/article_viewer/article_viewer.ts
class ArticleViewer
{
    #eventListenner(){
        
        $('.others-information .authors button').on('click', (event)=>{
            var spans = event.currentTarget.getElementsByTagName('span')
            var images = event.currentTarget.getElementsByTagName('img')
            if(event.currentTarget.previousElementSibling.getAttribute('data-show-less')==="true"){
                event.currentTarget.previousElementSibling.setAttribute('data-show-less', 'false')
                $(event.currentTarget.previousElementSibling).css('height', '100%')
                if(spans.length > 0)
                    spans[0].textContent = "View less"
                if(images.length > 0)
                    images[0].style.setProperty('rotate', '180deg')
                
            }
            else{
                event.currentTarget.previousElementSibling.setAttribute('data-show-less', 'true')
                $(event.currentTarget.previousElementSibling).css('height', '800px')

                if(spans.length > 0)
                    spans[0].textContent = "View more"
                if(images.length > 0)
                    images[0].style.setProperty('rotate', '0deg')
                
            }
        })
    }

    listen(){
        this.#eventListenner()
    }
}


class ArticleViewerActions
{
    isReading:boolean

    readDocument = (element:HTMLElement)  =>
    {
       var images = (element!.getElementsByTagName('img') as  HTMLCollectionOf<HTMLImageElement>)
       var span = element.nextElementSibling

       if(this.isReading == true)
        {
            if(images!.length > 0)
                images[0].src = "/static/images/cicle-filled-play-icon-2.webp"
            this.isReading = false

            location.href = '/editorial/'
        }
        else
        {
            if(images!.length > 0)
                images[0].src = "/static/images/cicle-filled-pause-icon-3.webp"
            this.isReading = true
        }
       
    }

    openEditorPage = (element:HTMLElement) =>
    {
        location.href = '/editorial/viewer'
    }

    goToPreviousPage = (element:HTMLElement)  =>
    {
        location.href = '/editorial/'
    }
}


var articleViewerActions = new ArticleViewerActions()

var articlesViewerFunctions = {
    "readDocument": articleViewerActions.readDocument,
    "previous": articleViewerActions.goToPreviousPage,
    "openEditor": articleViewerActions.openEditorPage
}


let a = (new ArticleViewer()).listen()